### PR TITLE
bash: default to 5

### PR DESCRIPTION
**Pull request by happysalada**
###### Motivation for this change

This is just a test to see if defaulting to bash 5 and removing old bash 4.4 works properly.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
@GrahamcOfBorg eval

**Comment by happysalada** [comment]
I've checked the error on my local, here is what I found out

```
local input=/nix/store/c9w69kl4caarcs2j84hck0icrdj9jqr6-fix-darwin-dylib-names.sh
++ local output=/nix/store/wiqpq8fwh190c9msc8iq68ai6cszca88-hook/nix-support/setup-hook
++ shift 2
++ '[' '!' -f /nix/store/c9w69kl4caarcs2j84hck0icrdj9jqr6-fix-darwin-dylib-names.sh ']'
++ local content
++ consumeEntire content
++ IFS=
++ read -r -N 0 content
++ echo 'consumeEntire(): ERROR: Input null bytes, won'\''t process'
consumeEntire(): ERROR: Input null bytes, won't process
++ return 1
```

the file
```
❯ cat /nix/store/c9w69kl4caarcs2j84hck0icrdj9jqr6-fix-darwin-dylib-names.sh
# On macOS, binaries refer to dynamic library dependencies using
# either relative paths (e.g. "libicudata.dylib", searched relative to
# $DYLD_LIBRARY_PATH) or absolute paths
# (e.g. "/nix/store/.../lib/libicudata.dylib").  In Nix, the latter is
# preferred since it allows programs to just work.  When linking
# against a library (e.g. "-licudata"), the linker uses the install
# name embedded in the dylib (which can be shown using "otool -D").
# Most packages create dylibs with absolute install names, but some do
# not.  This setup hook fixes dylibs by setting their install names to
# their absolute path (using "install_name_tool -id").  It also
# rewrites references in other dylibs to absolute paths.

fixupOutputHooks+=('fixDarwinDylibNamesIn $prefix')

fixDarwinDylibNames() {
    local flags=()
    local old_id

    for fn in "$@"; do
        flags+=(-change "$(basename "$fn")" "$fn")
    done

    for fn in "$@"; do
        if [ -L "$fn" ]; then continue; fi
        echo "$fn: fixing dylib"
        int_out=$(@targetPrefix@install_name_tool -id "$fn" "${flags[@]}" "$fn" 2>&1)
        result=$?
        if [ "$result" -ne 0 ] &&
            ! grep "shared library stub file and can't be changed" <<< "$out"
        then
            echo "$int_out" >&2
            exit "$result"
        fi
    done
}

fixDarwinDylibNamesIn() {
    local dir="$1"
    fixDarwinDylibNames $(find "$dir" -name "*.dylib")
}
```

**Comment by happysalada** [comment]
I've found 2 previous attempt and one person suggested a solution
https://github.com/NixOS/nixpkgs/pull/108101

second attempt is older and just like mine so far
https://github.com/NixOS/nixpkgs/pull/53687

I'm trying the fix documented in the first link. In bash 5 the behavior of `read -N 0` has changed and that was not documented anywhere. I'm testing this on local, I will push the commit after.

File: pkgs/stdenv/generic/setup.sh [hunk 1]
```diff
@@ -7,6 +7,8 @@ if [ -n "${BASH_VERSINFO-}" ] && [ "${BASH_VERSINFO-}" -lt 4 ]; then
     exit 1
 fi
 
+set -x
```
Comment (06kellyjac) [inline pkgs/stdenv/generic/setup.sh:10]: Isn't this covered by line 13?

Comment (happysalada) [reply pkgs/stdenv/generic/setup.sh:10]: that is a leftover from my debugging. Sorry for leaving that!
For some reason I can't get it to work with NIX_DEBUG, so I just forcefully added it.
My bad. Removing it from the commit

**Comment by happysalada** [comment]
The next two errors are test errors while building nix.
```
running test tests/user-envs.sh... [FAIL]
    + '[' -z '' ']'
    + clearStore
    + echo 'clearing store...'
    clearing store...
    + chmod -R +w /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + mkdir /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix
    + mkdir /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix
    + nix-store --init
    + clearProfiles
    + profiles=/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles
    + clearProfiles
    + profiles=/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles
    ++ nix-env -p /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles/test -q '*'
    ++ wc -l
    + test 0 -eq 0
    + mkdir -p /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/test-home
    + nix-env --switch-profile /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles/test
    ++ nix-env -f ./user-envs.nix -qa '*'
    ++ wc -l
    + test 6 -eq 6
    ++ nix-env -f ./user-envs.nix -qa --out-path --no-name '*'
    ++ grep foo-1.0
    + outPath10=/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/4rchg0azqmx526arwhv8a7b3817q1c3x-foo-1.0
    ++ nix-env -f ./user-envs.nix -qa --drv-path --no-name '*'
    ++ grep foo-1.0
    + drvPath10=/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/sq8ikm091c6jldai40sdp4ax420mab5f-foo-1.0.drv
    + '[' -n /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/4rchg0azqmx526arwhv8a7b3817q1c3x-foo-1.0 -a -n /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/sq8ikm091c6jldai40sdp4ax420mab5f-foo-1.0.drv ']'
    + nix-env -f ./user-envs.nix -qa '*' --description
    + grep -q silly
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/test-home/.nix-defexpr
    ++ pwd
    + ln -s /private/tmp/nix-build-nix-2.3.14.drv-0/nix-2.3.14/tests/user-envs.nix /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/test-home/.nix-defexpr
    + nix-env -qa '*' --description
    + grep -q silly
    + nix-env -qa '*' --system
    + grep -q x86_64-darwin
    + nix-env -i foo-1.0
    installing 'foo-1.0'
    these derivations will be built:
      /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/sq8ikm091c6jldai40sdp4ax420mab5f-foo-1.0.drv
    building '/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/sq8ikm091c6jldai40sdp4ax420mab5f-foo-1.0.drv'...
    error: unexpected EOF reading a line
```

```
running test tests/fetchurl.sh... [FAIL]
    + clearStore
    + echo 'clearing store...'
    clearing store...
    + chmod -R +w /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + mkdir /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix
    + mkdir /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix
    + nix-store --init
    + clearProfiles
    + profiles=/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles
    ++ nix-hash --flat --type sha256 ./fetchurl.sh
    + hash=9b1d81c5333e8c25296214cbf642e479131b8437f69d6e50cde80029a9e55e84
    +++ pwd
    ++ nix-build '<nix/fetchurl.nix>' --argstr url file:///private/tmp/nix-build-nix-2.3.14.drv-0/nix-2.3.14/tests/fetchurl.sh --argstr sha256 9b1d81c5333e8c25296214cbf642e479131b8437f69d6e50cde80029a9e55e84 --no-out-link --hashed-mirrors ''
    these derivations will be built:
      /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/jcmp6jplgqc3vdwj1pp3nxs5fj7gvqdk-fetchurl.sh.drv
    building '/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/jcmp6jplgqc3vdwj1pp3nxs5fj7gvqdk-fetchurl.sh.drv'...
    + outPath=/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/zr0g75mi67dazv4hby9lgj8jk14avh2v-fetchurl.sh
    + cmp /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/zr0g75mi67dazv4hby9lgj8jk14avh2v-fetchurl.sh fetchurl.sh
    + clearStore
    + echo 'clearing store...'
    clearing store...
    + chmod -R +w /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + mkdir /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix
    + mkdir /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix
    + nix-store --init
    + clearProfiles
    + profiles=/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles
    ++ nix hash-file --type sha512 --base64 ./fetchurl.sh
    + hash=aKjXWoXYIDfLkn4MV1P6FR/rkiXOqps6A4J3cWCYc1G3mgtTbIK5/YY7wXuExg/ltSf8JHlD25wCyovEU6ZI9Q==
    +++ pwd
    ++ nix-build '<nix/fetchurl.nix>' --argstr url file:///private/tmp/nix-build-nix-2.3.14.drv-0/nix-2.3.14/tests/fetchurl.sh --argstr sha512 aKjXWoXYIDfLkn4MV1P6FR/rkiXOqps6A4J3cWCYc1G3mgtTbIK5/YY7wXuExg/ltSf8JHlD25wCyovEU6ZI9Q== --no-out-link --hashed-mirrors ''
    these derivations will be built:
      /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/v4wknf7v874g819pr9vc20rw5inywb31-fetchurl.sh.drv
    building '/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/v4wknf7v874g819pr9vc20rw5inywb31-fetchurl.sh.drv'...
    + outPath=/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/qjwwfb7nq62w30pzmiijnl0289jfhnf7-fetchurl.sh
    + cmp /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/qjwwfb7nq62w30pzmiijnl0289jfhnf7-fetchurl.sh fetchurl.sh
    + clearStore
    + echo 'clearing store...'
    clearing store...
    + chmod -R +w /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + mkdir /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix
    + mkdir /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix
    + nix-store --init
    + clearProfiles
    + profiles=/private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles
    + rm -rf /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/var/nix/profiles
    ++ nix hash-file ./fetchurl.sh
    + hash=sha256-mx2BxTM+jCUpYhTL9kLkeRMbhDf2nW5QzegAKanlXoQ=
    + [[ sha256-mx2BxTM+jCUpYhTL9kLkeRMbhDf2nW5QzegAKanlXoQ= =~ ^sha256- ]]
    +++ pwd
    ++ nix-build '<nix/fetchurl.nix>' --argstr url file:///private/tmp/nix-build-nix-2.3.14.drv-0/nix-2.3.14/tests/fetchurl.sh --argstr hash sha256-mx2BxTM+jCUpYhTL9kLkeRMbhDf2nW5QzegAKanlXoQ= --no-out-link --hashed-mirrors ''
    these derivations will be built:
      /private/tmp/nix-build-nix-2.3.14.drv-0/nix-test/store/iwjcbsicdy9zbzr3mli6109c1lzf5zsv-fetchurl.sh.drv
    error: unexpected EOF reading a line
    + outPath=
```

it only took 19h on my machine to build.

**Comment by happysalada** [comment]
@vcunat when you have a moment, could we try a hydra rebuild, I would like to make sure this does not break stuff.
I understand you are busy, no worries if this takes time.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -10443,24 +10443,15 @@ in
 
   any-nix-shell = callPackage ../shells/any-nix-shell { };
 
-  bash = lowPrio (callPackage ../shells/bash/4.4.nix {
+  bash = lowPrio (callPackage ../shells/bash/5.1.nix {
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:10446]: Can we keep bash_4?

Comment (happysalada) [reply pkgs/top-level/all-packages.nix:10446]: Sure, what are you thinking about ? Just for people to be able to use?
under `bash4` then?

Comment (06kellyjac) [reply pkgs/top-level/all-packages.nix]: usually its `name_X` when extra versions are packaged for extra use (either older or newer than the main package)

**Comment by 06kellyjac** [comment]
I think you just rebased off master instead of staging :slightly_smiling_face:

**Comment by happysalada** [comment]
Yes :-)
I'm creating a new branch and then It shows i should not stay in front of the computer anymore.

**Comment by happysalada** [comment]
Recreated the PR leaving bash_4 around
https://github.com/NixOS/nixpkgs/pull/133495

Sorry about messing up the branch and notifying half of the world.

**Pull request closed without merging**